File: benchmarks/list-dot/bench.py
import time, struct
N = 1_000_000
def dot_product(n):
    a = [0.0] * n; b = [0.0] * n
    for i in range(n):
        v = (i % 100 + 1) / 100.0
        # Clamp to f32 precision
        v = struct.unpack('f', struct.pack('f', v))[0]
        a[i] = v; b[i] = struct.unpack('f', struct.pack('f', 1.0 - v))[0]
    return sum(a[i] * b[i] for i in range(n))
t = time.perf_counter()
r = dot_product(N)
print(f"f32 dot product (N={N}): {r:.6f} ({(time.perf_counter()-t)*1000:.1f}ms)")

File: benchmarks/list-dot/bench.rs
use std::time::Instant;
fn dot_product(n: usize) -> f32 {
    let mut a = Vec::with_capacity(n);
    let mut b = Vec::with_capacity(n);
    for i in 0..n {
        let v = ((i % 100 + 1) as f32) / 100.0f32;
        a.push(v);
        b.push(1.0f32 - v);
    }
    a.iter().zip(b.iter()).map(|(x, y)| x * y).sum()
}
fn main() {
    let n = 1_000_000usize;
    let t = Instant::now();
    let result = dot_product(n);
    println!("f32 dot product (N={}): {} ({}ms)", n, result, t.elapsed().as_millis());
}
